docs/understand/programming_model.rst

```diff
@@ -0,0 +1,158 @@
+
+  Block Diagram of an CDNA3 Compute Unit.
+
+For hardware implementation's sake, some number of Multi Processors are grouped
+together into a Shader Engine or Compute Engine, typically sharing some fixed
+function units or memory subsystem resources.
+
+.. _cdna2_gcd:
+
+.. figure:: ../data/understand/programming_model/cdna2_gcd.png
+  :alt: Block diagram showing four "Compute Engine"s each with 28 "CU"s
+        (Compute Unit) inside. These four Compute Engines share one block of
+        "L2 Cache and Controllers". Around them are four "Memory Controller"s,
+        each having a "Memory Phy" block next to them. To the top and bottom of
+        all these are eight blocks of "Infinity Fabric Link" with one of the
+        eight reading "Infinity Fabric or PCIe". Two sole "VCN" blocks sit in
+        top corners. At the very bottom spans a colored seaction reading
+        "Infinity Fabric" along with its logo.
+
+  Block Diagram of a CDNA2 Graphics Compute Die.
+
+Single Instruction Multiple Threads
+===============================================================================
+
+The SIMT programming model behind the HIP device-side execution is a
+middle-ground between SMT (Simultaneous Multi-Threading) programming known from
+multi-core CPUs, and SIMD (Single Instruction, Multiple Data) programming
+mostly known from exploiting relevant instruction sets on CPUs (eg.
+SSE/AVX/Neon).
+
+A HIP device compiler maps our SIMT code written in HIP C++ to an inherently
+SIMD architecture (like GPUs) not by exploiting data parallelism within a
+single instance of a kernel and spreading identical instructions over the SIMD
+engines at hand, but by scalarizing the entire kernel and issuing the scalar
+instructions of multiple kernel instances to each of the SIMD engine lanes.
+
+Consider the following kernel
+
+.. code:: cu
+
+  __global__ void k(float4* a, const float4* b)
+  {
+    int tid = threadIdx.x;
+    int bid = blockIdx.x;
+    int dim = blockDim.x;
+  
+    a[tid] += (tid + bid - dim) * b[tid];
+  }
+
+The incoming four-vector of floating-point values ``a`` is multiplied by a
+scalar and then multiplied element-wise by another four-vector. On modern
+SIMD-capable architectures the four-vector ops are expected to compile to a
+single SIMD instruction. GPU execution of this kernel however will typically
+look the following:
+
+.. _simt:
+
+.. figure:: ../data/understand/programming_model/simt.svg
+  :alt: Two large arrows pointing downward with blocks inside and ellipses
+        between the arrows. Inside the arrows the same series of blocks with
+        the following texts inside from top to bottom: "ADD", "DIV", "FMA",
+        "FMA", "FMA" and "FMA".
+
+  Instruction flow of the sample SIMT program.
+
+In HIP, lanes of a SIMD architecture are fed by mapping threads of a SIMT
+execution, one thread down each lane of a SIMD engine. Execution parallelism
+isn't exploited from the width of the built-in vector types, but via the thread
```

Comment:
This is where presenting ILP and TLP at the same time can become complicated - there are vector instructions, that can also be executed in parallel

Reply:
Slightly reworded